File: src/Entities/Wrappers/EntityMoverFollow.cs
using Celeste.Mod.Entities;
using FMOD.Studio;
using Microsoft.Xna.Framework;
using Monocle;
using System;

namespace Celeste.Mod.GameHelper.Entities.Wrappers;

[CustomEntity("GameHelper/EntityMoverFollow")]
public class EntityMoverFollow : EntityMover {
    private readonly bool onlyX, onlyY, stopSoundOnLeave, holdPositionOnWait, awaitPlayerMovement;
    private readonly string flag, approachMode, approachSound, targetOnlyType;
    private readonly float speedFactor, playSoundAtDistance, minDistance, maxDistance;
    private readonly Vector2 offset;
    private Entity moveTarget;
    private Vector2 waitPosition;
    private EventInstance approachSoundEvent;
    private bool soundPlayed;

    public EntityMoverFollow(EntityData data, Vector2 levelOffset) : base(data, levelOffset) {
        approachMode = data.Attr("approachMode");
        speedFactor = data.Float("speedFactor");
        flag = data.Attr("flag");
        onlyX = data.Bool("onlyX");
        onlyY = data.Bool("onlyY");
        targetOnlyType = data.Attr("targetOnlyType");
        approachSound = data.Attr("approachSound");
        playSoundAtDistance = data.Float("playSoundAtDistance");
        stopSoundOnLeave = data.Bool("stopSoundOnLeave");
        holdPositionOnWait = data.Bool("holdPositionOnWait");
        awaitPlayerMovement = data.Bool("awaitPlayerMovement");
        minDistance = data.Float("minDistance");
        maxDistance = data.Float("maxDistance");
        offset = new(data.Float("offsetX"), data.Float("offsetY"));
    }

    public override void Update() {
        base.Update();

        float dist = (moveTarget.Position + offset - target.Position).Length();
        Player p = SceneAs<Level>().Tracker.GetEntity<Player>();
        if(
            (maxDistance > 0 && dist > maxDistance) ||
            dist <= minDistance ||
            !Utils.Util.GetFlag(flag, Scene, true) ||
            (p != null && awaitPlayerMovement && p.JustRespawned)
        ) {
            soundPlayed = false;
            if(holdPositionOnWait) {
                moveTo(waitPosition);
            }
            if(stopSoundOnLeave && dist > minDistance && Audio.IsPlaying(approachSoundEvent)) {
                Audio.Stop(approachSoundEvent);
            }
            return;
        }

        Vector2 targetPos = Calc.Approach(
            target.Position,
            moveTarget.Position + offset,
            approach(dist, approachMode, speedFactor)
        );
        if(onlyX ^ onlyY) {
            if(onlyX) {
                moveTo(new Vector2(targetPos.X, target.Position.Y));
            } else {
                moveTo(new Vector2(target.Position.X, targetPos.Y));
            }
        } else {
            moveTo(targetPos);
        }
        waitPosition = target.Position;

        if(approachSound.Length == 0) return;

        if((moveTarget.Position + offset - target.Position).Length() <= playSoundAtDistance) {
            if(!soundPlayed) {
                soundPlayed = true;
                approachSoundEvent = Audio.Play(approachSound);
            }
        } else {
            soundPlayed = false;
            if(Audio.IsPlaying(approachSoundEvent)) {
                Audio.Stop(approachSoundEvent);
            }
        }
    }

    private static float approach(float dist, string modus, float factor) {
        if(modus.StartsWith("Hyperbolic") || modus.StartsWith("Inverse")) {
            dist = Math.Max(dist, 1f);
        }
        dist /= 60f;
        return modus switch {
            "Exponential" => (float) Math.Pow(2, dist * factor),
            "Cubic" => factor * dist * dist * dist,
            "Quadratic" => factor * dist * dist,
            "Linear" => factor * dist,
            "Constant" => factor,
            "Hyperbolic" => factor / dist,
            "Inverse Quadratic" => factor / dist / dist,
            "Inverse Cubic" => factor / dist / dist / dist,
            "Inverse Exponential" => 1 / (float) Math.Pow(2, dist / factor),
            _ => 0,
        };
    }

    public override void Awake(Scene scene) {
        target = FindNearest(nodes[0], onlyType);
        moveTarget = FindNearest(nodes[1], targetOnlyType, target);
        if(target == null || moveTarget == null) {
            Logger.Warn("GameHelper",
                "Follow Entity Mover found " +
                (target == null ? "no entity" : "entity " + target.GetType().ToString()) + ", " +
                (moveTarget == null ? "no target" : "target " + moveTarget.GetType().ToString()) +
                " in room " + SceneAs<Level>().Session.LevelData.Name
            );
            RemoveSelf();
            return;
        }
        if(debug) {
            Logger.Info("GameHelper", "Follow Entity Mover found entity " + target.GetType().ToString() + " and target " + moveTarget.GetType().ToString());
        }
        waitPosition = target.Position;
        base.Awake(scene);
    }

    public override void Removed(Scene scene) {
        base.Removed(scene);
        if(stopSoundOnLeave && Audio.IsPlaying(approachSoundEvent)) {
            Audio.Stop(approachSoundEvent);
        }
    }
}

File: Loenn/entities/entity_mover_follow.lua
local mover = {}

local approachModi = {
    a_Exponential = "Exponential",
    b_Cubic = "Cubic",
    c_Quadratic = "Quadratic",
    d_Linear = "Linear",
    e_Constant = "Constant",
    f_Hyperbolic = "Hyperbolic",
    g_Inverse_Quadratic = "Inverse Quadratic",
    h_Inverse_Cubic = "Inverse Cubic",
    i_Inverse_Exponential = "Inverse Exponential"
}

mover.name = "GameHelper/EntityMoverFollow"
mover.depth = -999999999
mover.texture = "loenn/GameHelper/entity_mover_follow"
mover.justification = {0, 0}
mover.nodeLimits = {1, 1}
mover.nodeLineRenderType = "line"
mover.placements = {
    name = "follow",
    data = {
        approachMode = "Constant",
        speedFactor = 1,
        flag = "",
        onlyX = false,
        onlyY = false,
        naiveMovement = false,
        onlyType = "",
        targetOnlyType = "",
        approachSound = "",
        playSoundAtDistance = 100.0,
        stopSoundOnLeave = false,
        holdPositionOnWait = false,
        debug = false,
        awaitPlayerMovement = true,
        minDistance = 0.0,
        maxDistance = 0.0,
        offsetX = 0.0,
        offsetY = 0.0
    }
}
mover.fieldOrder = {
    "x", "y",
    "approachMode", "speedFactor",  "onlyType", "targetOnlyType", "minDistance", "maxDistance",
    "offsetX", "offsetY", "approachSound", "playSoundAtDistance", "flag",
    "onlyX", "onlyY", "naiveMovement", "holdPositionOnWait","awaitPlayerMovement", "stopSoundOnLeave", "debug"
}
mover.fieldInformation = {
    approachMode = {
        options = approachModi,
        editable = false
    }
}

return mover